### PR TITLE
fix: bound macro expansion depth across body and block boundaries

### LINKED ISSUE #22840: stack overflow: recursive macro
```rust
#[macro_export]
macro_rules! test {
    () => {
        #[test]
        fn my_test() {
            test!();
        }
    };
}

#[cfg(test)]
mod tests {
    use super::*;
    test!();
}

fn main() {}
```

`~/.rustup/toolchains/master/bin/rust-analyzer analysis-stats --run-all-ide-things`

```
Database loaded:     1.31s, 224minstr, 48mb (metadata 638.95ms, 923kinstr, 150kb; build 201.70ms, 165kinstr, 1088b)
  item trees: 1
  dependency lines of code: 1_462_633, item trees: 2_900
  dependency item stats: traits: 851, impl: 9_701, mods: 3_012, macro calls: 5_501, macro rules: 1_053
Item Tree Collection: 9.51s, 22ginstr, 425mb
  Total Statistics:
    crates: 1, mods: 2, decls: 4, bodies: 2, adts: 0, consts: 0, signatures: 2, variants: 0
  Workspace:
    traits: 0, macro_rules macros: 1, proc_macros: 0
    lines of code: 17, item trees: 1
    usages: traits: 0, impl: 0, mods: 1, macro calls: 1, macro rules: 1
  Dependencies:
    lines of code: 1_462_633, item trees: 2_900
    declarations: traits: 851, impl: 9_701, mods: 3_012, macro calls: 5_501, macro rules: 1_053
Item Collection:     5.26s, 11ginstr, 177mb
Crate lang items:    1.02s, 2472minstr, 31mb
Expression Store Lowering: 622.02µs, 382kinstr, 163kb
  exprs: 2, ??ty: 0 (0%), ?ty: 0 (0%), !ty: 0
  pats: 0, ??ty: 0 (100%), ?ty: 0 (100%), !ty: 0
  panics: 0
Inference:           3.18ms, 766kinstr, 2487kb
MIR lowering:        152.60µs, 95kinstr, 944b
Mir failed bodies: 0 (0%)
Data layouts:        250.00ns, 24kinstr, 0b
Failed data layouts: 0 (100%)
Const evaluation:    46.93µs, 48kinstr, 368b
Failed const evals: 0 (100%)
0/1 0% diagnostics: /tmp/ra/a/src/main.rs
thread 'main' (620753) has overflowed its stack
fatal runtime error: stack overflow, aborting
[1]    620753 abort      RUST_BACKTRACE=full ~/.rustup/toolchains/master/bin/rust-analyzer   .
```

### PR DESCRIPTION
**Pull request by rit3sh-x**
A macro that expands to an item invoking the macro again overflows the stack:

```rust
macro_rules! test {
    () => { fn my_test() { test!(); } };
}
mod tests { test!(); }
```

rustc rejects this with "recursion limit reached while expanding `test!`". We build an infinite tree of nested items instead, and `DefWithBody::diagnostics` walks it until the stack runs out.

The recursion limit never fires because there are two expansion depth counters and both restart from zero at every body and every block, `Expander::new` and `DefCollector::seed_with_inner`. Each expansion of `test!` advances its counter exactly one level before the next body starts a fresh count, so neither ever reaches the limit.

Fixes https://github.com/rust-lang/rust-analyzer/issues/22840

### CODE REVIEW

File: crates/hir-def/src/lib.rs [hunk 1]
```diff
@@ -109,6 +109,34 @@ pub(crate) struct ExpandProcAttrMacros {
...
+/// an item that invokes the macro again therefore advances neither counter past one, never trips
+/// the limit, and builds an infinite tree of nested items. Seeding those counters from here
+/// restores rustc's semantics.
+///
+/// The walk stops once the depth exceeds `cap`, as callers only compare it against their
+/// recursion limit. That keeps this O(`cap`) rather than O(chain length).
+pub(crate) fn macro_expansion_depth(
+    db: &dyn SourceDatabase,
+    mut file: HirFileId,
+    cap: usize,
```
Comment (Veykril) [inline crates/hir-def/src/lib.rs]: let's replace all the `usize` we use for this logic here and in callers with u32, thats what the limits work with simplifying the code a bit by dropping casts

File: crates/hir-def/src/expr_store/expander.rs [hunk 2]
```diff
@@ -16,15 +16,18 @@ use tt::TextRange;
...
 
 #[derive(Debug)]
 pub(super) struct Expander<'db> {
     span_map: SpanMap<'db>,
     current_file_id: HirFileId,
     ast_id_map: &'db AstIdMap,
-    /// `recursion_depth == usize::MAX` indicates that the recursion limit has been reached.
+    /// Counts from the depth `current_file_id` already sits at, so that the budget is shared by
+    /// the whole expansion chain rather than restarting for every body.
+    ///
```
Comment (Veykril) [inline crates/hir-def/src/expr_store/expander.rs]:
```suggestion
```
Unrelevant info, that is explained by the source already

File: crates/ide-diagnostics/src/handlers/macro_error.rs [hunk 3]
```diff
@@ -328,6 +328,56 @@ fn it_works() {
         );
     }
 
+    // A macro that expands to an item re-invoking the macro produces an infinite tree of nested
+    // items: the expansion depth counters reset at every body and every block boundary, so the
+    // recursion limit never trips and walking the tree overflows the stack. rustc rejects this
+    // program with "recursion limit reached while expanding `test!`".
+    //
+    // `recursion_limit` is lowered so the test stays cheap. Note the `cfg!(test)` cap in
+    // `Expander::new` does not apply here — hir-def is a dependency, not the crate under test.
```
Comment (Veykril) [inline crates/ide-diagnostics/src/handlers/macro_error.rs]:
```suggestion
```

File: crates/ide-diagnostics/src/handlers/macro_error.rs [hunk 4]
```diff
@@ -328,6 +328,56 @@ fn it_works() {
...
+}
+mod tests {
+    test!();
+}
+"#,
+        );
+    }
+
+    // The same shape reached through `#[macro_export]` and `$crate`, as in the issue — the
+    // expansion chain then runs through a different macro definition path.
```
Comment (Veykril) [inline crates/ide-diagnostics/src/handlers/macro_error.rs]:
```suggestion
```

**Review by Veykril — COMMENTED** [review]
lgtm minus most of the comment junk

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-analyzer/22974/e4e67338eaf4724e4e6e1525711e00b85d174cf9..7f786947ef003a0e3c620a9408fe61ba1ec98787)*

**Comment by rit3sh-x** [comment]
@Veykril can you review again?

**Comment by ChayimFriedman2** [comment]
First, if you used AI to author this patch, please note that [our AI policy](https://github.com/rust-lang/rust-analyzer/blob/master/AI_POLICY.md) requires disclosure.

Now to the code: before merging this, I want to see what the speed impact on `analysis-stats` is. Please check not only on r-a itself as it doesn't use macros a lot. We commonly test on https://github.com/facebook/buck2/ and https://github.com/oxidecomputer/omicron.

**Comment by rit3sh-x** [comment]
> First, if you used AI to author this patch, please note that [our AI policy](https://github.com/rust-lang/rust-analyzer/blob/master/AI_POLICY.md) requires disclosure.
> 
> Now to the code: before merging this, I want to see what the speed impact on `analysis-stats` is. Please check not only on r-a itself as it doesn't use macros a lot. We commonly test on https://github.com/facebook/buck2/ and https://github.com/oxidecomputer/omicron.

I used AI for making different edge cases for testing out the implementation.

**Comment by rit3sh-x** [comment]
@ChayimFriedman2 @Veykril I tested the buck2 repository over my implementation against base. This is the bench result:

| Metric | base | PR | Δ | base spread over 3 runs |
|---|---:|---:|---:|---|
| crate def map time | 44691 ms | 45441 ms | +1.7% | 41656–50548 |
| body lowering time | 11547 ms | 10972 ms | −5.0% | 10983–11783 |
| item tree time | 16663 ms | 16569 ms | −0.6% | 16161–17895 |
| total time | 73233 ms | 74405 ms | +1.6% | 70231–81110 |
| total memory | 5119 MB | 5232 MB | +2.2% | 5109–5282 |

Do you think this can be optimised?

**Comment by ChayimFriedman2** [comment]
Speed is within noise, but memory... Adding 113mb is not something to do lightly. We'll do it if we have no other choice since the outcome here is pretty bad even though the code is invalid, but let me first think if there is another way.

**Comment by rit3sh-x** [comment]
> Speed is within noise, but memory... Adding 113mb is not something to do lightly. We'll do it if we have no other choice since the outcome here is pretty bad even though the code is invalid, but let me first think if there is another way.

Could you try doing this benchmark on your machine? my machine has a bit less RAM, so some data is skewed due to use of SWAP memory.

**Comment by rit3sh-x** [comment]
Also, I ran PR variants first so there might be some TLB cache involved as well.

**Comment by ChayimFriedman2** [comment]
> Could you try doing this benchmark on your machine? my machine has a bit less RAM, so some data is skewed due to use of SWAP memory.

I will.

**Comment by ChayimFriedman2** [comment]
Okay, after working quite a lot to get a reliable memory benchmark (I switched OS recently and the new OS had much larger variance), I was able to get a very deterministic and reliable benchmark, and I'm happy to tell you that this change does not impact performance considerably (a few more megabytes, that's all).

I did manage to create a slightly more efficient version (~1mb less, ~1-2 seconds less). If you want I can push it to this PR, but if you do not want (e.g. because you want the PR to be all yours) I agree not to.

**Comment by rit3sh-x** [comment]
@ChayimFriedman2 could you guide me how can I optimise this?

**Comment by ChayimFriedman2** [comment]
I can push it here, but if you want to do it yourself, the basic idea is: instead of having a separate query, store the macro depth in the `MacroCallId` when interning it, as it's already available then.

**Comment by rit3sh-x** [comment]
@ChayimFriedman2 I pushed the change, could you review it.

File: crates/hir-expand/src/lib.rs [hunk 5]
```diff
@@ -1099,6 +1100,19 @@ impl MacroDefId {
...
+    pub(crate) fn new(
+        db: &dyn SourceDatabase,
+        def: MacroDefId,
+        krate: Crate,
+        kind: MacroCallKind,
+        ctxt: SyntaxContext,
+    ) -> MacroCallLoc {
+        // The parent's depth is a field read on an already interned loc, so this is O(1) and the
+        // whole chain stays linear.
+        let depth = kind.file_id().macro_expansion_depth(db).saturating_add(1);
```
Comment (ChayimFriedman2) [inline crates/hir-expand/src/lib.rs]: In all places creating a `MacroCallId` the depth is already available. If we go by this route, we should forward it and not read it from the query.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-analyzer/22974/e4e67338eaf4724e4e6e1525711e00b85d174cf9..5f924b0fc6c9c0fc145c1bfc6a8885642f76f529)*

**Review by ChayimFriedman2 — CHANGES_REQUESTED** [review]

**Comment by rit3sh-x** [comment]
@ChayimFriedman2 Does this look right? `depth` is now only the recursion budget, and `call_depth` is the canonical expansion depth.

File: crates/hir-def/src/attrs/docs.rs [hunk 6]
```diff
@@ -378,6 +378,7 @@ fn expand_doc_macro_call<'db>(
             expander.resolver.resolve_path_as_macro_def(expander.db, path, Some(MacroSubNs::Bang))
         },
         &mut |_, _| (),
+        ast_id.file_id.macro_expansion_depth(expander.db).saturating_add(1),
```
Comment (ChayimFriedman2) [inline crates/hir-def/src/attrs/docs.rs]: Here too the depth is available.

File: crates/hir-def/src/expr_store/expander.rs [hunk 7]
```diff
@@ -113,6 +113,9 @@ impl<'db> Expander<'db> {
                 krate,
                 |path| resolver(path).map(|it| it.definition(db)),
                 eager_callback,
+                // `recursion_depth` is the depth of `current_file_id`, which is where this call
+                // is written; the expansion sits one level below it.
+                this.recursion_depth.saturating_add(1),
```
Comment (ChayimFriedman2) [inline crates/hir-def/src/expr_store/expander.rs]: All the `saturating_add()` could and should be replaced by `+`; reaching `u32::MAX` is impossible due to the recursion limit.

File: crates/hir-def/src/nameres/assoc.rs [hunk 8]
```diff
@@ -325,6 +325,7 @@ impl<'db> AssocItemCollector<'db> {
                     &mut |ptr, call_id| {
                         self.macro_calls.push((ptr.map(|(_, it)| it.upcast()), call_id))
                     },
+                    ast_id.file_id.macro_expansion_depth(self.db).saturating_add(1),
```
Comment (ChayimFriedman2) [inline crates/hir-def/src/nameres/assoc.rs]: And again the depth is available.

File: crates/hir-def/src/nameres/attr_resolution.rs [hunk 9]
```diff
@@ -73,6 +73,7 @@ impl DefMap {
             AttrMacroAttrIds::from_one(attr_id),
             self.krate,
             def.definition(db),
+            ast_id.ast_id.file_id.macro_expansion_depth(db).saturating_add(1),
```
Comment (ChayimFriedman2) [inline crates/hir-def/src/nameres/attr_resolution.rs]: Here too, as I said in all places it is available.

**Review by ChayimFriedman2 — CHANGES_REQUESTED** [review]
Frankly I don't see the point of trying to get it to what I already have. If you want, I can push what I have. If not, let's just merge the old version.

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-analyzer/22974/e4e67338eaf4724e4e6e1525711e00b85d174cf9..c393458d3ca3394d185c9b4b6344d92d9b0153fa)*

**Comment by rit3sh-x** [comment]
@ChayimFriedman2 after these changes will this be ready for merge? I just wanted to optimise it 😅.

**Comment by ChayimFriedman2** [comment]
If I push my changes? Yes. If we decide not to push them and revert to the old version? Also yes.

**Comment by rit3sh-x** [comment]
@ChayimFriedman2 I made one commit can you review if this is not good we can merge yours or revert to prev commit as you see fit.

**Review by ChayimFriedman2 — APPROVED** [review]
That was still not perfect, I pushed mine like you said.

Thank you!

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-analyzer/22974/36896409dc6dbf741f487b84fc084cdf0e7876e6..ddbe6aca6ca283f5e8cfa78a2549b4472ec1b684)*
